controller/grpc.go

```diff
@@ -998,9 +1006,73 @@ outer:
 	return res, nil
 }
 
-func maybeError(err error) error {
-	if err == nil {
-		return nil
+func (g *grpcAPI) ListCertificates(ctx context.Context, req *api.ListCertificatesRequest) (*api.ListCertificatesResponse, error) {
+	certs, err := g.routeRepo.ListCertificates()
+	if err != nil {
+		return nil, err
+	}
+	apiCerts := make([]*api.Certificate, len(certs))
+	for i, cert := range certs {
+		apiCerts[i] = api.NewCertificate(cert)
 	}
-	return api.NewError(err, err.Error())
+	return &api.ListCertificatesResponse{
+		Certificates: apiCerts,
+	}, nil
+}
+
+func (g *grpcAPI) GetCertificate(ctx context.Context, req *api.GetCertificateRequest) (*api.GetCertificateResponse, error) {
+	cert, err := g.routeRepo.GetCertificate(req.Name)
+	if err != nil {
+		return nil, err
+	}
+	return &api.GetCertificateResponse{
+		Certificate: api.NewCertificate(cert),
+	}, nil
+}
+
+func (g *grpcAPI) CreateCertificate(ctx context.Context, req *api.CreateCertificateRequest) (*api.CreateCertificateResponse, error) {
+	// expect a static certificate
+	v, ok := req.Certificate.(*api.CreateCertificateRequest_Static_)
+	if !ok {
+		return nil, status.Error(codes.InvalidArgument, "only static certificates are supported")
+	}
+	staticCert := v.Static
+
+	// PEM encode the key and certificate chain
+	var certPEM bytes.Buffer
+	for _, certBytes := range staticCert.Chain {
+		if err := pem.Encode(&certPEM, &pem.Block{Type: "CERTIFICATE", Bytes: certBytes}); err != nil {
+			return nil, err
+		}
+	}
+	var keyPEM bytes.Buffer
+	if err := pem.Encode(&keyPEM, &pem.Block{Type: "PRIVATE KEY", Bytes: staticCert.PrivateKey}); err != nil {
```

Comment:
The certificate chain shouldn't be PEM-encoded in the database, it should be stored as an array of DER binary certificates. The key should also just be DER binary.

Reply:
Done in 3bf6384b.